WIP Off Canvas Navigation: Edit Button

georgeh opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Partially addresses #45442

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This will simplify editing menu items from the navigation inspector

## How?
This is based off of the (closed) #44534 branch and will be rebased once the follow-up PRs get merged. As of e1b012a83b the only branch-specific changes are in `packages/block-editor/src/components/off-canvas-editor/block-edit-button.js`

- [x] Add edit botton
- [x] Trigger convert-to-links when editing an auto menu
- [ ] ~~Trigger block inspector overlay for other navigation child blocks~~ moving to follow-up PR
- [ ] Trigger convert-to-links when editing a link under the auto menu

<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

## Screenshots or screencast <!-- if applicable -->

draganescu commented:
Cool stuff 🙇🏻 this is the very next step needed before the drill down system can be added to the experiment.

<details>
<summary>This adds an edit button to the list view. Right now when I click on it I get the dialogue for converting a non existent page list block to navigation links 🐳  </summary>


https://user-images.githubusercontent.com/107534/200594062-3c25481a-5e61-402c-b6dd-62cf9094dbfc.mp4



</details>

That dialogue should only appear if there is a page list block used in the navigation menu and the user wants to edit a link in the page list. Otherwise when clicking the edit button a panel with the detail selected appears (a la global styles).

But the icon, positioning and interaction feels very neat! 👏🏻

georgeh commented:
@draganescu 
> That dialogue should only appear if there is a page list block used in the navigation menu and the user wants to edit a link in the page list. Otherwise when clicking the edit button a panel with the detail selected appears (a la global styles).

Thanks for the clarification, I'll be sure to update it. One edge case I found was that we only offer the edit button on Page List blocks with fewer than 100 children:

https://github.com/WordPress/gutenberg/blob/6f0e046c4119636a6dc419481d56f4509e7e7b08/packages/block-library/src/page-list/edit.js#L26-L28

My thought is to disable the button with an explanation in that situation, but that's a lower priority.

That's what I'm working on next. I'm going to try to reuse the InspectorPanel slots for the slide-over. In the video I noticed that the Page Link edit example is a bit different from the InspectorPanel.

<img width="243" alt="Screenshot showing off-canvas editing with option to add URL" src="https://user-images.githubusercontent.com/128240/200622774-4198e8de-6a46-4e40-bd5c-14138671f920.png">

Right now I am thinking we can use InspectorPanel initially and possibly create another slot for off-canvas editing to add things like the URL field.

draganescu commented:
Excellent. One thing that would be good is to explore the sliding UI in a new PR for clarity and ease of review. 

Also apparently all my PRs for the list view have been merged and so you cans base your work  on trunk now.

SaxonF commented:
One note for the "back" interaction is that we should build it in a way that can apply to any scenario. e.g. if block is passed as prop to inspector show back button that navigates back to that block

Commit:
Rebased: Initial work on nav edit button

It was easier to nuke the commit history than to manually sync a bunch of commits,
so this is all the commits in the branch squashed

georgeh commented:
@draganescu I squashed my commits to rebase off of trunk, so the diff should be much more usable. Good call on splitting up the sliding panel part from this, it'll be much easier to land this with the fixes for the Page List and follow up with the edit panel for individual items.

Commit:
Lint

Commit:
"Fix" unit tests

draganescu commented on packages/block-editor/package.json:
```diff
@@ -48,6 +49,7 @@
 		"@wordpress/html-entities": "file:../html-entities",
 		"@wordpress/i18n": "file:../i18n",
 		"@wordpress/icons": "file:../icons",
+		"@wordpress/interface": "file:../interface",
```
What is this for?

georgeh replied:
Legacy from the sliding panel stuff, I'm surprised the linter didn't complain. I'll pull that out since I'm going to be doing the panel in a follow-up

Commit:
Remove extra dependency

georgeh commented:
> That dialogue should only appear if there is a page list block used in the navigation menu and the user wants to edit a link in the page list. Otherwise when clicking the edit button a panel with the detail selected appears (a la global styles).

@draganescu I'm not seeing links underneath the page list. Is that something that should be working? Or should I add that to this PR?

<img width="1440" alt="Screenshot 2022-11-10 at 9 42 48 AM" src="https://user-images.githubusercontent.com/128240/201155140-71de43e8-f23b-4913-a4db-1119a5711ebe.png">

Commit:
Hide edit button from non-automenus

Commit:
Lint

draganescu commented:
Page list does not show its contents in the list view. That is a feature of the page list block that should be added in the future - in a PR specifically for that as it's about that block not about the navigation block.

scruffian commented:
Connected to this, I created a PR to add the URL field to the Navigation Link inspector controls panel: https://github.com/WordPress/gutenberg/pull/45751

georgeh commented:
@draganescu We've moved the panels out of scope for this PR, are there any more features we need before we merge this?

getdave commented on packages/block-editor/package.json:
```diff
@@ -39,6 +39,7 @@
 		"@wordpress/blocks": "file:../blocks",
 		"@wordpress/components": "file:../components",
 		"@wordpress/compose": "file:../compose",
+		"@wordpress/core-data": "file:../core-data",
```
The block editor shouldn't have core data as a dependency as Core Data is WP specific whereas block editor is platform agnostic.

getdave commented on packages/block-editor/src/components/off-canvas-editor/block-edit-button.js:
```diff
@@ -0,0 +1,225 @@
+	};
+
+const ConvertToLinksModal = ( { onClose, clientId } ) => {
+	const { records: pages, hasResolved: pagesFinished } = useEntityRecords(
```
We will need to find another way to grab the Pages records from Core.

Likely you will need to pass them as part of the block editor settings via the relevant `edit/*` store. This way the same code could be utilised across the block and this off canvas list view.

Here is an example of using `@wordpress/editor` to get information about Core and pass it into the block editor settings for retrieval thereby avoiding the need for the `@wordpress/core-data` dep in the `@wordpress/block-editor` package.



https://github.com/WordPress/gutenberg/blob/eb3faf6ffac764ef9b7293750c45fa94d1d6c348/packages/editor/src/components/provider/use-block-editor-settings.js#L60

https://github.com/WordPress/gutenberg/blob/eb3faf6ffac764ef9b7293750c45fa94d1d6c348/packages/editor/src/components/provider/use-block-editor-settings.js#L190

georgeh replied:
Thanks, I'll see how I can get the data we need into the component without calling `core-data` directly.

getdave reviewed (changes requested):
Previous review should have been blocking to ensure we don't accidentally commit `@wordpress/core-data` as a dependency of `@wordpress/block-editor`.

See my previous review for details.

getdave commented:
@georgeh I'm working on removing the `core-data` dependency from this PR.

getdave committed:
Ongoing work to decouple core-data

getdave commented:
Ok I suggest we down tools on this one until we're absolutely clear that 

- the "edit" icon on the parent Page List will be required
- the "edit" button will be required

@draganescu is working on making the child items of Page List visible in List View so that may invalidate the need for this PR.

getdave committed:
Fetch pages via block editor settings

getdave committed:
Revert additional of core-data dependency

getdave committed:
Revert core data dep in block editor package

getdave commented on packages/block-editor/src/components/off-canvas-editor/block-edit-button.js:
```diff
@@ -0,0 +1,215 @@
+	);
+};
+
+const BlockEditButton = ( { label, clientId } ) => {
```
As things stand I'm pretty sure this code is going to run for all types of blocks. It's actually only required for the Page List block. It will need to be conditionalised.

getdave commented:
FYI I decided a good first step would be to land the edit button on it's own. Therefore I spun out https://github.com/WordPress/gutenberg/pull/45815.

I've put @georgeh down as co-author on that PR as much of the work was borrowed from this one.

In the future we _may_ need to implement the work from this PR around Auto Menu conversion. Hopefully not as it adds complexity but at least we have it available should we require it.
